pkg/controller/gkenetworkparamset/gkenetworkparamset_controller.go

```diff
@@ -0,0 +1,135 @@
+
+	"go.uber.org/multierr"
+	networkv1alpha1 "k8s.io/cloud-provider-gcp/crd/apis/network/v1alpha1"
+	networkclientset "k8s.io/cloud-provider-gcp/crd/client/network/clientset/versioned"
+	networkinformers "k8s.io/cloud-provider-gcp/crd/client/network/informers/externalversions/network/v1alpha1"
+	"k8s.io/cloud-provider-gcp/pkg/util/gkenetworkparamsetutil"
+	"k8s.io/cloud-provider-gcp/providers/gce"
+	controllersmetrics "k8s.io/component-base/metrics/prometheus/controllers"
+)
+
+// Controller manages GKENetworkParamSet status.
+type Controller struct {
+	gkeNetworkParamsInformer networkinformers.GKENetworkParamSetInformer
+	networkClientset         networkclientset.Interface
+	informerStartFn          func(stopCh <-chan struct{})
+	gceCloud                 *gce.Cloud
+}
+
+// NewGKENetworkParamSetController returns a new
+func NewGKENetworkParamSetController(
+	networkClientset networkclientset.Interface,
+	gkeNetworkParamsInformer networkinformers.GKENetworkParamSetInformer,
+	informerStartFn func(stopCh <-chan struct{}),
+	gceCloud *gce.Cloud,
+) *Controller {
+
+	return &Controller{
+		networkClientset:         networkClientset,
+		gkeNetworkParamsInformer: gkeNetworkParamsInformer,
+		informerStartFn:          informerStartFn,
+		gceCloud:                 gceCloud,
+	}
+
+}
+
+// Run starts an asynchronous loop that monitors and updates GKENetworkParamSet in the cluster.
+func (c *Controller) Run(stopCh <-chan struct{}, controllerManagerMetrics *controllersmetrics.ControllerManagerMetrics) {
+	defer utilruntime.HandleCrash()
+
+	klog.Infof("Starting gkenetworkparamset controller")
+	defer klog.Infof("Shutting down gkenetworkparamset controller")
+	controllerManagerMetrics.ControllerStarted("gkenetworkparamset")
+	defer controllerManagerMetrics.ControllerStopped("gkenetworkparamset")
+
+	c.gkeNetworkParamsInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    gkenetworkparamsetutil.CreateAddGKENetworkParamSetHandler(c.onGKENetworkParamSetAdded),
+		UpdateFunc: gkenetworkparamsetutil.CreateUpdateGKENetworkParamSetHandler(c.onGKENetworkParamSetUpdated),
+	})
+
+	c.informerStartFn(stopCh)
+
+	<-stopCh
+	klog.Infof("done gkenetworkparamset controller")
+}
+
+func (c *Controller) onGKENetworkParamSetAdded(params *networkv1alpha1.GKENetworkParamSet) error {
+	return c.updateGKENetworkParamSet(params)
+}
+func (c *Controller) onGKENetworkParamSetUpdated(oldParams, newParams *networkv1alpha1.GKENetworkParamSet) error {
+	return c.updateGKENetworkParamSet(newParams)
+}
+
+func (c *Controller) updateGKENetworkParamSet(params *networkv1alpha1.GKENetworkParamSet) (rerr error) {
+	ctx := context.Background()
+	cidrs := []string{}
+
+	defer func() {
+		err := gkenetworkparamsetutil.UpdateGKENetworkParamSetStatus(ctx, c.networkClientset.NetworkingV1alpha1().GKENetworkParamSets(), params, cidrs)
+		if err != nil {
+			klog.Error(err, "failed to patch GKENetworkParamSetStatus")
+			rerr = multierr.Append(rerr, err)
```

Comment:
this is complex to read, if we can avoid the named returns parameters the better, I think we can avoid the defer per comment https://github.com/kubernetes/cloud-provider-gcp/pull/417/commits/8776c6b97afebc7c5cf863db5b18878470702c87#r1067492964

In case we need it avoid the external dependency on multierr, and use golang wrapping return fmt.Errorf("failed to update: %v, err)